nestadia/src/lib.rs

```diff
@@ -116,6 +126,10 @@ impl Emulator {
         self.cartridge.get_save_data()
     }
 
+    pub fn take_audio_samples(&mut self) -> alloc::vec::Drain<i16> {
```

Comment:
I don't think using a `Drain` here is effective and might be part of the performance issue. Wouldn't it be better to take ownership of the `Vec` directly instead?

Reply:
I wonder... From what I read, `Drain` moves the content of the `Vec` while `IntoIter` moves the whole `Vec` (as an iterator). Because the `Vec` is being reused, I thought the Drain was more appropriate, because otherwise I would have to reallocate a new one everytime the samples are taken.

`IntoIter` probably wouldn't be any better, because it's still an iterator. Only way to move that I can see is replacing the vec by an `Option<Vec>`, having to check at every clock that it's not none (or assume it is with an unwrap), and to take the samples call the `take` method of the `Option` to move and set to None (and possibly reallocate right there to not have to check for none anywhere else). Is that really faster though?

There's also another issue, in libretro I use `vec.map()` to duplicate the mono samples to stereo, which means there won't be much gain there as it's still an iterator... unless I guess I convert the sample vec to stereo right away by pushing the sample twice

Reply:
So I tried moving the vec, and I get "cannot move out of `self.samples` which is behind a mutable reference" (I this case `&mut self`)